online/src/main/scala/ai/chronon/online/OnlineDerivationUtil.scala

```diff
@@ -0,0 +1,128 @@
+
+import ai.chronon.aggregator.windowing.TsUtils
+import ai.chronon.api.Extensions.DerivationOps
+import ai.chronon.api.{Derivation, LongType, StringType, StructField, StructType}
+import ai.chronon.online.Fetcher.Request
+
+object OnlineDerivationUtil {
+  type DerivationFunc = (Map[String, Any], Map[String, Any]) => Map[String, Any]
+
+  val timeFields: Array[StructField] = Array(
+    StructField("ts", LongType),
+    StructField("ds", StringType)
+  )
+
+  // remove value fields of groupBys that have failed with exceptions
+  // and reintroduce the exceptions back
+  private[online] def reintroduceExceptions(derived: Map[String, Any],
+                                            preDerivation: Map[String, Any]): Map[String, Any] = {
+    val exceptions: Map[String, Any] = preDerivation.iterator.filter(_._1.endsWith("_exception")).toMap
+    if (exceptions.isEmpty) {
+      return derived
+    }
+    val exceptionParts: Array[String] = exceptions.keys.map(_.dropRight("_exception".length)).toArray
+    derived.filterKeys(key => !exceptionParts.exists(key.startsWith)).toMap ++ exceptions
+  }
+
+  def buildRenameOnlyDerivationFunction(derivationsScala: List[Derivation]): DerivationFunc = {
+    {
+      case (_: Map[String, Any], values: Map[String, Any]) =>
+        reintroduceExceptions(derivationsScala.applyRenameOnlyDerivation(values), values)
+    }
+  }
+
+  private def buildDerivationFunctionWithSql(
+      catalystUtil: PooledCatalystUtil
+  ): DerivationFunc = {
+    {
+      case (keys: Map[String, Any], values: Map[String, Any]) =>
+        reintroduceExceptions(catalystUtil.performSql(keys ++ values).orNull, values)
+    }
+  }
+
+  def buildDerivationFunction(
+      derivationsScala: List[Derivation],
+      keySchema: StructType,
+      baseValueSchema: StructType
+  ): DerivationFunc = {
+    if (derivationsScala.isEmpty) {
+      return { case (_, values: Map[String, Any]) => values }
+    } else if (derivationsScala.areDerivationsRenameOnly) {
+      buildRenameOnlyDerivationFunction(derivationsScala)
+    } else {
+      val catalystUtil = buildCatalystUtil(derivationsScala, keySchema, baseValueSchema)
+      buildDerivationFunctionWithSql(catalystUtil)
+    }
+  }
+
+  def applyDeriveFunc(
+      deriveFunc: DerivationFunc,
+      request: Request,
+      baseMap: Map[String, AnyRef]
+  ): Map[String, AnyRef] = {
+    val requestTs = request.atMillis.getOrElse(System.currentTimeMillis())
+    val requestDs = TsUtils.toStr(requestTs).substring(0, 10)
+    // used for derivation based on ts/ds
+    val tsDsMap: Map[String, AnyRef] = {
+      Map("ts" -> (requestTs).asInstanceOf[AnyRef], "ds" -> (requestDs).asInstanceOf[AnyRef])
+    }
+    val derivedMap: Map[String, AnyRef] = Try(
+      deriveFunc(request.keys, baseMap ++ tsDsMap)
+        .mapValues(_.asInstanceOf[AnyRef])
+        .toMap) match {
+      case Success(derivedMap) => derivedMap
+      case Failure(exception) => {
+        throw exception
```

Comment:
let's not throw here. We could use datadog's increment exception method to capture it.